File: clients/go/consensus/tx_dep_graph.go
package consensus

import (
	"bytes"
	"sort"
)

// TxDepEdge represents a dependency between two non-coinbase transactions
// within the same block. ProducerIdx and ConsumerIdx are 0-based indices
// into the non-coinbase transaction slice (i.e. pb.Txs[1:]).
type TxDepEdge struct {
	ProducerIdx int
	ConsumerIdx int
	Kind        TxDepEdgeKind
}

// TxDepEdgeKind classifies dependency edges.
type TxDepEdgeKind uint8

const (
	// DepParentChild means ConsumerIdx spends an output created by ProducerIdx
	// within the same block.
	DepParentChild TxDepEdgeKind = iota

	// DepSamePrevout means both transactions attempt to spend the same
	// pre-existing UTXO. In a valid block only one can succeed; the graph
	// records the ordering constraint so that parallel validation can
	// detect the conflict deterministically.
	DepSamePrevout
)

// TxDepGraph is a deterministic dependency graph over non-coinbase transactions
// in a single block. It supports topological-level assignment for parallel
// validation scheduling.
type TxDepGraph struct {
	// TxCount is the number of non-coinbase transactions.
	TxCount int

	// Edges is the complete list of dependency edges, sorted deterministically.
	Edges []TxDepEdge

	// Levels assigns each transaction to a topological level (0-based).
	// Transactions at the same level have no inter-dependencies and may be
	// validated in parallel. Level 0 transactions depend only on the
	// block-start UTXO snapshot.
	Levels []int

	// LevelOrder is a deterministic ordering of transaction indices, grouped
	// by level. Within each level, transactions are sorted by their txid
	// (lexicographic, ascending) for deterministic tie-breaking.
	LevelOrder []int

	// MaxLevel is the highest level in the graph (0-based).
	MaxLevel int
}

// BuildTxDepGraph constructs a deterministic dependency graph from precomputed
// transaction contexts. The contexts slice corresponds to non-coinbase
// transactions (pb.Txs[1:]).
//
// Two types of edges are detected:
//  1. Parent-child: tx[j] spends an output created by tx[i] where i < j.
//     Detected via txid match between input outpoints and earlier tx outputs.
//  2. Same-prevout: tx[i] and tx[j] both spend the same outpoint from the
//     block-start UTXO snapshot. The lower-index tx is the "producer" for
//     ordering purposes.
//
// After edge construction, the graph computes topological levels using the
// longest-path algorithm, then produces a deterministic LevelOrder using
// lexicographic txid tie-breaking within each level.
func BuildTxDepGraph(contexts []TxValidationContext) *TxDepGraph {
	n := len(contexts)
	if n == 0 {
		return &TxDepGraph{}
	}

	// Map txid → context index for parent-child detection.
	txidToIdx := make(map[[32]byte]int, n)
	for i := range contexts {
		txidToIdx[contexts[i].Txid] = i
	}

	// Map outpoint → first consumer index for same-prevout detection.
	outpointFirstConsumer := make(map[Outpoint]int)

	var edges []TxDepEdge

	for i := range contexts {
		for _, op := range contexts[i].InputOutpoints {
			// Parent-child: does this outpoint reference an earlier tx in the block?
			if producerIdx, ok := txidToIdx[op.Txid]; ok && producerIdx < i {
				edges = append(edges, TxDepEdge{
					ProducerIdx: producerIdx,
					ConsumerIdx: i,
					Kind:        DepParentChild,
				})
				continue
			}

			// Same-prevout: another tx already consumed this outpoint.
			if firstIdx, ok := outpointFirstConsumer[op]; ok && firstIdx != i {
				// Lower index is always "producer" for determinism.
				lo, hi := firstIdx, i
				if lo > hi {
					lo, hi = hi, lo
				}
				edges = append(edges, TxDepEdge{
					ProducerIdx: lo,
					ConsumerIdx: hi,
					Kind:        DepSamePrevout,
				})
			} else if !ok {
				outpointFirstConsumer[op] = i
			}
		}
	}

	// Sort edges deterministically: by ConsumerIdx, then ProducerIdx, then Kind.
	sort.Slice(edges, func(a, b int) bool {
		if edges[a].ConsumerIdx != edges[b].ConsumerIdx {
			return edges[a].ConsumerIdx < edges[b].ConsumerIdx
		}
		if edges[a].ProducerIdx != edges[b].ProducerIdx {
			return edges[a].ProducerIdx < edges[b].ProducerIdx
		}
		return edges[a].Kind < edges[b].Kind
	})

	// Compute topological levels using longest-path from roots.
	levels := make([]int, n)
	for _, e := range edges {
		if levels[e.ProducerIdx]+1 > levels[e.ConsumerIdx] {
			levels[e.ConsumerIdx] = levels[e.ProducerIdx] + 1
		}
	}

	maxLevel := 0
	for _, l := range levels {
		if l > maxLevel {
			maxLevel = l
		}
	}

	// Build level order with lexicographic txid tie-break.
	order := make([]int, n)
	for i := range order {
		order[i] = i
	}
	sort.Slice(order, func(a, b int) bool {
		la, lb := levels[order[a]], levels[order[b]]
		if la != lb {
			return la < lb
		}
		return bytes.Compare(contexts[order[a]].Txid[:], contexts[order[b]].Txid[:]) < 0
	})

	return &TxDepGraph{
		TxCount:    n,
		Edges:      edges,
		Levels:     levels,
		LevelOrder: order,
		MaxLevel:   maxLevel,
	}
}

File: clients/rust/fuzz/fuzz_targets/tx_dep_graph.rs
#![no_main]
use libfuzzer_sys::fuzz_target;

fuzz_target!(|data: &[u8]| {
    // Cap input to avoid OOM.
    if data.len() > 8192 || data.is_empty() {
        return;
    }

    // Decode tx_count from first byte: 1–8 transactions.
    let tx_count = (data[0] % 8) as usize + 1;
    let rest = &data[1..];

    // Build TxValidationContexts from fuzz data, mirroring Go FuzzTxDepGraphBuild.
    let mut contexts = Vec::with_capacity(tx_count);
    for i in 0..tx_count {
        let mut txid = [0u8; 32];
        txid[0] = i as u8;

        let mut input_outpoints = Vec::new();

        // Parent-child edge: reference a previous tx's output.
        if i > 0 && rest.len() > i {
            let prev_idx = rest[i] as usize % i;
            let mut prev_txid = [0u8; 32];
            prev_txid[0] = prev_idx as u8;
            input_outpoints.push(rubin_consensus::Outpoint {
                txid: prev_txid,
                vout: 0,
            });
        }

        // Same-prevout conflict: reference an external outpoint.
        let edge_b_offset = tx_count;
        if rest.len() > edge_b_offset + i {
            let mut ext_txid = [0u8; 32];
            ext_txid[0] = 0xFF;
            ext_txid[1] = rest[edge_b_offset + i];
            input_outpoints.push(rubin_consensus::Outpoint {
                txid: ext_txid,
                vout: 0,
            });
        }

        contexts.push(rubin_consensus::TxValidationContext {
            tx_index: i + 1, // 1-based, coinbase excluded
            txid,
            input_outpoints,
        });
    }

    let graph = rubin_consensus::build_tx_dep_graph(&contexts);

    // Invariant: TxCount matches input.
    assert_eq!(
        graph.tx_count, tx_count,
        "tx_count mismatch: got {} want {}",
        graph.tx_count, tx_count
    );

    // Invariant: Levels length matches TxCount.
    assert_eq!(
        graph.levels.len(),
        tx_count,
        "levels length mismatch"
    );

    // Invariant: all levels are non-negative and ≤ max_level.
    for (idx, &lvl) in graph.levels.iter().enumerate() {
        assert!(
            lvl <= graph.max_level,
            "level {} exceeds max_level {} at index {}",
            lvl,
            graph.max_level,
            idx
        );
    }

    // Invariant: level_order is a valid permutation of [0..tx_count).
    assert_eq!(
        graph.level_order.len(),
        tx_count,
        "level_order length mismatch"
    );
    let mut seen = vec![false; tx_count];
    for &idx in &graph.level_order {
        assert!(
            idx < tx_count,
            "level_order index out of range: {}",
            idx
        );
        assert!(!seen[idx], "level_order duplicate: {}", idx);
        seen[idx] = true;
    }

    // Invariant: level_order is sorted by level (non-decreasing).
    for w in graph.level_order.windows(2) {
        assert!(
            graph.levels[w[0]] <= graph.levels[w[1]],
            "level_order not sorted by level"
        );
    }

    // Invariant: edges respect level ordering (producer < consumer in levels).
    for edge in &graph.edges {
        assert!(
            graph.levels[edge.producer_idx] < graph.levels[edge.consumer_idx],
            "edge level violation: producer level {} >= consumer level {}",
            graph.levels[edge.producer_idx],
            graph.levels[edge.consumer_idx]
        );
    }

    // Determinism: second call must produce identical graph.
    let graph2 = rubin_consensus::build_tx_dep_graph(&contexts);
    assert_eq!(graph, graph2, "non-deterministic graph");
});
